### DISCUSSION TITLE
ResolveFaceCollisions() is O(n²)

### DISCUSSION THREAD

**Discussion by srett** [photoprism/photoprism#1872] [Q&A]
Today I tried naming all the faces that are still marked "new" and don't belong to any cluster via the API. Apparently that generates a new entry in the faces table for each instance, even though I only have ~80 subjects. After a while I noticed that photoprism is constantly hogging one CPU core, which I eventually tracked down to `ResolveFaceCollisions`, which compares every face to every other face. My faces table currently has 1535 entries, and the function takes 9 minutes to run. I stopped adding any more faces at this point as this would get much worse quickly. :-)
At a quick glance, pulling the [check](https://github.com/photoprism/photoprism/blob/develop/internal/query/faces.go#L177) for `f1.SubjUID == f2.SubjUID` up, before matching the embeddings, might speed things up a bit, as well as making the inner loop only start running from the current index of the outer loop. But with a big enough collection, O(n²) will always eventually be a problem. Never the less, I'll test if this small change makes an impact and report back.

Right now this is probably not much of an issue yet, as nobody is going to tag 1500 faces by hand. I actually just stumbled upon this when I tried importing known face assignments from DigiKam. But as soon as PP has official means of importing faces from photos or sidecar files, this might become more of a problem.

See you all in 2022 :+1:

**Comment by lastzero** [comment]
Somewhat expected as the API is a first version for use with the UI. Can be optimized for other use cases, in particular for batch edits and if processing can be delayed and no instant feedback is required.

**Comment by srett** [comment]
Problem is that the function is called by the maintenance worker, every 15 minutes by default. I've increased this to 12 hours for now.

**Comment by lastzero** [comment]
Yes, but doesn't it get to a point where conflicts are resolved? Then maybe there are ambiguities. Use the audit command.

**Comment by srett** [comment]
There are no conflicts anymore. It's just that 1500 faces means 2.2M comparisons via `Face::Match()`, i.e. deserializing the JSON data from embeddings etc. At 2000 faces that would already mean 4M comparisons, so almost twice as many. Rough estimates are, if I would finish naming all faces, I'll have about 7000 faces in that table. That would be 49M comparisons, or a runtime of 3 hours.

A solution would be to introduce a `checked_at` column and then for the outer loop, only query faces where `checked_at < updated_at`, and maybe throw in a `LIMIT 100` for good measure. The inner loop still needs to iterate over all faces, but it should already be *much* faster.  
A stop-gap solution without fiddling with the DB scheme is to pick two random sets of let's say 250 faces and compare those. It should eventually resolve all conflicts, but obviously not in a deterministic time frame. I'm currently running this locally and switched the background worker back to 15 mins. `ResolveFaceCollisions()` now takes only about 10 seconds.

**Comment by lastzero** [comment]
The faces table contains FACE CLUSTERS. If you add multiple similar faces with core 1, they will be replaced with a single face cluster that has a radius that covers both. However, that process is asynchronous. And since you increased the interval, it happens only every 12 hours now. Happy New Year!

**Comment by srett** [comment]
Right, Happy New Year to you too :-)

Is it possible that the clustering doesn't work for faces that haven't been assigned to a cluster at all in the first place? Or is it just converging very slowly?  
What I mean is that I have no more clusters shown in the "New" Tab under "People". I already assigned the proper names to every cluster that ever popped up there. I only see faces when I click "Show all new faces". So if the clustering of unknown faces was not possible in the first place, maybe it also fails after assigning the proper Subject to all those faces?

When I noticed the issue yesterday, I first set the log level to debug and investigated further, so the worker ran at least 3 or 4 more times, still the amount of faces in the table never decreased, but then again I didn't take note of the exact number of rows, just remembered that it was around 1500. I will restore the original binary again now and let it run for a few more hours to see if the number of rows in the faces table changes.

**Comment by lastzero** [comment]
It's the wrong strategy for your use case. Creating too many clusters and then trying to merge them again is very expensive as you noticed. It's a tradeoff so users get instant feedback in the UI. There are other ways, but describing them in detail takes a while... Not working today 🎉

**Comment by srett** [comment]
No worries, was just using the holidays to fiddle around a bit, trying out if it was feasible to get all my digiKam data into PhotoPrism that way. Since it has pretty complete and manually reviewed face tags (have been using it for ages), it seemed worth a try.

**Comment by lastzero** [comment]
It might work like this if you add the names really slow, so the background worker can run in between. In this case, you also would expect many faces to be automatically recognized, so that there's no need to explicitly set a name in the first place. For batch imports, you would update the markers table only and then let the worker do its magic. The REST API doesn't offer this though.

**Comment by lastzero** [comment]
Shall we move this to discussions?

**Comment by srett** [comment]
Might be for the best; I'm planning to run a few more experiments to get this going, so this is getting more off-topic as I keep reporting back. Maybe it will help someone else trying to migrate.

It's probably still worth keeping the O(n²) thing in the back of your head, it's [bitten other people before](https://randomascii.wordpress.com/2021/02/16/arranging-invisible-icons-in-quadratic-time/). :-)

**Comment by lastzero** [comment]
Yes, that's why we do our best to keep the number of different faces small. Automatically creating a "cluster" for every single face doesn't scale and isn't how it's supposed to work.
